cyto_dl/models/im2im/diffusion_autoencoder.py

```diff
@@ -0,0 +1,227 @@
+        save_images_every_n_epochs: int
+            Image saving frequency
+        batch_is_cond: bool
+            Whether the batch is a tensor of latent features for conditioning the diffusion model. Only used during inference. If True, a latent walk is performed, otherwise latent features are extracted from the condition image using the semantic encoder. 
+        n_noise_samples: Optional[int]
+            Number of noise samples to average for latent walk
+        train_encoder: bool
+            Whether to train the semantic encoder
+        **base_kwargs:
+            Additional arguments passed to BaseModel
+        """
+        
+        _DEFAULT_METRICS = {
+            "train/loss": MeanMetric(),
+            "val/loss": MeanMetric(),
+            "test/loss": MeanMetric(),
+        }
+
+        metrics = base_kwargs.pop("metrics", _DEFAULT_METRICS)
+        super().__init__(metrics=metrics, **base_kwargs)
+        self.diffusion_key = diffusion_key or condition_key
+        self.automatic_optimization = True
+        self.autoencoder = autoencoder
+        if not train_encoder:
+            for param in semantic_encoder.parameters():
+                param.requires_grad = False
+        self.semantic_encoder = semantic_encoder
+
+        self.scheduler = DDIMScheduler(n_train_steps)
+        self.inferer = DiffusionInferer(self.scheduler)
+        self.spatial_inferer = spatial_inferer
+
+        self.loss = torch.nn.MSELoss()
+
+    def configure_optimizers(self):
+        params = list(self.autoencoder.parameters())
+        if self.hparams.train_encoder:
+            params += list(self.semantic_encoder.parameters())
+
+        opt = self.optimizer["generator"](params)
+        sched = self.lr_scheduler["generator"](optimizer=opt)
+        return [opt], [sched]
+
+    def forward(self, x_cond, x_diff):
+        noise = torch.randn_like(x_diff, device=x_diff.device)
+        timesteps = torch.randint(
+            0,
+            self.inferer.scheduler.num_train_timesteps,
+            (x_diff.shape[0],),
+            device=x_diff.device,
+            dtype=torch.long,
+        )
+        # latent is B x C x 1
+        latent = self.semantic_encoder(x_cond).unsqueeze(2)
+        noise_pred = self.inferer(
+            inputs=x_diff,
+            diffusion_model=self.autoencoder,
+            noise=noise,
+            timesteps=timesteps,
+            condition=latent,
+        )
+        return noise, noise_pred, latent
+
+    def _generate_image(self, noise, cond):
+        with torch.no_grad():
+            sample, intermediates = self.inferer.sample(
+                input_noise=noise,
+                diffusion_model=self.autoencoder,
+                scheduler=self.scheduler,
+                save_intermediates=True,
+                conditioning=cond,
+                verbose=False,
+                intermediate_steps=1,
+            )
+        # during training, final image is all-nan, try to return last non-nan image
```

Comment:
why is this?

Reply:
not sure.... it goes away during inference